csharp/src/Ice/ObjectAdapter.cs

```diff
@@ -876,36 +878,106 @@ public async Task SetPublishedEndpointsAsync(IEnumerable<Endpoint> newEndpoints)
+                if (!current.IsOneway)
+                {
+                    response.Finish();
+                    dispatchObserver?.Reply(response.Size);
+                }
+                return response;
+            }
+            catch (Exception ex)
+            {
+                if (!current.IsOneway)
+                {
+                    RemoteException actualEx;
+                    if (ex is RemoteException remoteEx && !remoteEx.ConvertToUnhandled)
+                    {
+                        actualEx = remoteEx;
+                        dispatchObserver?.RemoteException();
+                    }
+                    else
+                    {
+                        actualEx = new UnhandledException(current.Identity, current.Facet, current.Operation, ex);
+                        dispatchObserver?.Failed(actualEx.InnerException!.GetType().FullName ?? "System.Exception");
+                    }
+
+                    if (Communicator.WarnDispatch && actualEx is UnhandledException)
+                    {
+                        Warning(ex);
+                    }
+                    var response = new OutgoingResponseFrame(request, actualEx);
+                    response.Finish();
+                    dispatchObserver?.Reply(response.Size);
+                    return response;
+                }
+                else
+                {
+                    if (Communicator.WarnDispatch)
+                    {
+                        Warning(ex);
+                    }
+                    dispatchObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
+                    return OutgoingResponseFrame.WithVoidReturnValue(current);
+                }
+            }
+            finally
+            {
+                dispatchObserver?.Detach();
+            }
 
-            ValueTask<OutgoingResponseFrame> DispatchAsync(
-                IncomingRequestFrame request,
-                Current current,
-                IObject servant,
-                int i)
+            ValueTask<OutgoingResponseFrame> DispatchAsync(IObject servant, int i)
             {
                 if (i < Interceptors.Count)
                 {
                     DispatchInterceptor interceptor = Interceptors[i++];
 
                     return interceptor(request,
                                        current,
-                                       (request, current) => DispatchAsync(request, current, servant, i));
+                                       (request, current) => DispatchAsync(servant, i));
                 }
                 else
                 {
                     return servant.DispatchAsync(request, current);
                 }
             }
+
+            void Warning(Exception ex)
+            {
+                var output = new StringBuilder();
+                output.Append("dispatch exception:");
+                output.Append("\nidentity: ").Append(current.Identity.ToString(Communicator.ToStringMode));
```

Comment:
We need to rework this ice1-style printing (for another PR?).

Reply:
I can fix it but what's the new style?